exercises.py

```diff
@@ -1,28 +1,40 @@
+from typing import Dict
+
 """
 Recursion Problem Set: Coding Exercises
 """
 
 def reverse(s: str) -> str:
-    """
-    Reverse string using recursive approach
-    """
+	if len(s) == 1:
+		return s
+	else:
+		return reverse(s[1:]) + reverse(s[0])
 
 def fib_rec(n: int) -> int:
-    """
-    Recursive fibonacci
-    """
+	if n < 2:
+		return n
+	else:
+		return fib_rec(n - 2) + fib_rec(n - 1)
 
 def fib_dyn(n: int) -> int:
-    """
-    Fibonacci using dynamic programming (w/ a cache)
-    """
+	cache: Dict[int, int] = {0: 0, 1: 1}
```

Comment:
⛔️  We need our cache to be outside of the function. Otherwise, each call to `fib_dyn` will have its on locally scoped `cache`. Love the typing though.